### PR TITLE
[SYCL] Refactor context to enhance ABI compatibility

### PR DESCRIPTION
**Pull request by alexbatashev**
This patch is a part of effort to decouple SYCL Runtime library
interface from its actual implementation. The goal is to improve
SYCL ABI/API compatibility between different versions of library.

The following modifications were applied to context and context_impl
classes:

1. Removed include of context_impl header from context.hpp and replaced
it with forward declaration.
2. std::shared_ptr was replaced with shared_ptr_class for context class
3. Documentation was improved for both context and context_impl
4. context_impl now tries to follow LLVM code style

Applying aforementioned changes requires modifying other header files.
While some of those may seem as a regression, affected classes will be
covered by future patches.

Signed-off-by: Alexander Batashev <alexander.batashev@intel.com>

### CODE REVIEW

File: sycl/include/CL/sycl/detail/context_impl.hpp [hunk 1]
```diff
@@ -28,55 +27,110 @@ class device;
...
   std::map<RT::PiProgram, std::map<string_class, RT::PiKernel>> &
   getCachedKernels() {
-    return m_CachedKernels;
+    return MCachedKernels;
   }
 
+  /// Gets USM dispatch.
+  ///
+  /// @return a pointer to USM dispatch.
   std::shared_ptr<usm::USMDispatcher> getUSMDispatch() const;
```
Comment (alexbatashev) [inline sycl/include/CL/sycl/detail/context_impl.hpp]: TODO: this 3 methods need better wording for documentation

Comment (alexbatashev) [reply sycl/include/CL/sycl/detail/context_impl.hpp]: Actually, I could not come up with a better description, so any suggestions are welcome

File: sycl/include/CL/sycl/detail/queue_impl.hpp [hunk 2]
```diff
@@ -9,6 +9,7 @@
 #pragma once
 
 #include <CL/sycl/context.hpp>
+#include <CL/sycl/detail/context_impl.hpp>
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/detail/queue_impl.hpp]: Why do you include this header?

File: sycl/include/CL/sycl/detail/program_impl.hpp [hunk 3]
```diff
@@ -10,6 +10,7 @@
 #include <CL/sycl/context.hpp>
 #include <CL/sycl/detail/common.hpp>
 #include <CL/sycl/detail/common_info.hpp>
+#include <CL/sycl/detail/context_impl.hpp>
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/detail/program_impl.hpp]: Why do you include this header?

File: sycl/include/CL/sycl/detail/sycl_mem_obj_t.hpp [hunk 4]
```diff
@@ -8,8 +8,8 @@
 
 #pragma once
 
-#include <CL/sycl/context.hpp>
 #include <CL/sycl/detail/common.hpp>
+#include <CL/sycl/detail/context_impl.hpp>
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/detail/sycl_mem_obj_t.hpp]: Why?

File: sycl/source/context.cpp [hunk 5]
```diff
@@ -23,59 +23,51 @@
...
+context::context(const platform &Platform, async_handler AsyncHandler)
+    : context(Platform.get_devices(), AsyncHandler) {}
 
-context::context(const vector_class<device> &deviceList,
-                 async_handler asyncHandler) {
-  if (deviceList.empty()) {
+context::context(const vector_class<device> &DeviceList,
+                 async_handler AsyncHandler) {
+  if (DeviceList.empty()) {
     throw invalid_parameter_error("First argument deviceList is empty.");
```
Comment (romanovvlad) [inline sycl/source/context.cpp]:
```suggestion
    throw invalid_parameter_error("Device list is empty.");
```

File: sycl/include/CL/sycl/context.hpp [hunk 6]
```diff
@@ -46,20 +95,34 @@ class context {
...
   bool is_host() const;
 
+  /// Gets platform associated with this SYCL context.
+  ///
+  /// @return a valid instance of SYCL platform.
   platform get_platform() const;
 
+  /// Gets devices associated with this SYCL context.
+  ///
+  /// @return a list of valid SYCL device instances.
```
Comment (vladimirlaz) [inline sycl/include/CL/sycl/context.hpp]: vector is more exact wording.

File: sycl/source/detail/sampler_impl.cpp [hunk 7]
```diff
@@ -6,7 +6,7 @@
 //
 //===----------------------------------------------------------------------===//
 
-#include <CL/sycl/context.hpp>
+#include <CL/sycl/detail/context_impl.hpp>
```
Comment (vladimirlaz) [inline sycl/source/detail/sampler_impl.cpp]: only context class is used. Can we have farward declaration here?

File: sycl/source/detail/scheduler/commands.cpp [hunk 8]
```diff
@@ -11,6 +11,7 @@
 #include "CL/sycl/access/access.hpp"
 #include <CL/cl.h>
 #include <CL/sycl/detail/clusm.hpp>
+#include <CL/sycl/detail/context_impl.hpp>
```
Comment (vladimirlaz) [inline sycl/source/detail/scheduler/commands.cpp]: forward declaration?

File: sycl/include/CL/sycl/detail/context_impl.hpp [hunk 9]
```diff
@@ -28,55 +27,110 @@ class device;
...
-  async_handler m_AsyncHandler;
-  vector_class<device> m_Devices;
-  RT::PiContext m_Context;
-  platform m_Platform;
-  bool m_OpenCLInterop;
-  bool m_HostContext;
-  std::map<OSModuleHandle, RT::PiProgram> m_CachedPrograms;
-  std::map<RT::PiProgram, std::map<string_class, RT::PiKernel>> m_CachedKernels;
-  std::shared_ptr<usm::USMDispatcher> m_USMDispatch;
+  async_handler MAsyncHandler;
```
Comment (vladimirlaz) [inline sycl/include/CL/sycl/detail/context_impl.hpp]: In LLVM coding style M prefix is not needed for class fields.

File: sycl/source/detail/sampler_impl.cpp (hunk 7 above)
Comment (alexbatashev) [reply sycl/source/detail/sampler_impl.cpp]: @vladimirlaz see https://github.com/intel/llvm/pull/873/files/55ec3d0c56e7555c720f9bdf190ec160739a584c#diff-ff508afff1be2b73d2e5cece4232ccc5R60

```
getSyclObjImpl(Context)->getHandleRef()
```
is an instance of `context_impl`, so we do need it.

File: sycl/source/detail/scheduler/commands.cpp (hunk 8 above)
Comment (alexbatashev) [reply sycl/source/detail/scheduler/commands.cpp]: Same here:
```
detail::getSyclObjImpl(MQueue->get_context())
```

File: sycl/include/CL/sycl/detail/context_impl.hpp (hunk 9 above)
Comment (alexbatashev) [reply sycl/include/CL/sycl/detail/context_impl.hpp]: I agree it is not necessary to have `M` prefix, but I think code is more readable when you can easily distinguish class fields and local variables.

File: sycl/include/CL/sycl/detail/sycl_mem_obj_t.hpp (hunk 4 above)
Comment (alexbatashev) [reply sycl/include/CL/sycl/detail/sycl_mem_obj_t.hpp]: @romanovvlad you're right, no need here

File: sycl/include/CL/sycl/detail/program_impl.hpp (hunk 3 above)
Comment (alexbatashev) [reply sycl/include/CL/sycl/detail/program_impl.hpp]:
```
detail::getSyclObjImpl(Context)->getHandleRef()
```
in line 94, for example

File: sycl/include/CL/sycl/detail/queue_impl.hpp (hunk 2 above)
Comment (alexbatashev) [reply sycl/include/CL/sycl/detail/queue_impl.hpp]:
```
RT::PiContext Context = detail::getSyclObjImpl(m_Context)->getHandleRef();
```
in line 162 for example

**Pull request merged**